### PR TITLE
Allow Nix mode to be disabled under NixOS

### LINKED ISSUE #4002: stack does not honor --no-nix under NixOS
Commit b60819cbe9dc5e97cdc3771a0fb538d21ab190d2 introduces the “Note: Enabling Nix integration, as it is required under NixOS”. Sometimes you *don't* want to use nix mode, especially when using the docker integration at the same time for producing binaries for other architectures.

### Steps to reproduce
run stack --no-nix under NixOS, watch the message “Note: Enabling Nix integration, as it is required under NixOS”.

### Expected
Don't use nix integration when --no-nix is set

It's real bad. Right now I have to revert since the build toolchain is broken because of stack.

### Version
```
Version 1.7.1 x86_64
Compiled with:
- Cabal-2.2.0.1
- Glob-0.9.2
- HUnit-1.6.0.0
- QuickCheck-2.10.1
- StateVar-1.1.1.0
- aeson-1.2.4.0
- aeson-compat-0.3.7.1
- annotated-wl-pprint-0.7.0
- ansi-terminal-0.8.0.2
- ansi-wl-pprint-0.6.8.2
- array-0.5.2.0
- asn1-encoding-0.9.5
- asn1-parse-0.9.4
- asn1-types-0.3.2
- async-2.1.1.1
- attoparsec-0.13.2.2
- attoparsec-iso8601-1.0.0.0
- auto-update-0.1.4
- base-4.10.1.0
- base-compat-0.9.3
- base-orphans-0.7
- base-prelude-1.2.0.1
- base16-bytestring-0.1.1.6
- base64-bytestring-1.0.0.1
- basement-0.0.7
- bifunctors-5.5.2
- binary-0.8.5.1
- bindings-uname-0.1
- bitarray-0.0.1.1
- blaze-builder-0.4.1.0
- blaze-html-0.9.0.1
- blaze-markup-0.8.2.1
- byteable-0.1.1
- bytestring-0.10.8.2
- call-stack-0.1.0
- case-insensitive-1.2.0.11
- cereal-0.5.5.0
- clock-0.7.2
- colour-2.3.4
- comonad-5.0.3
- conduit-1.3.0.2
- conduit-extra-1.3.0
- connection-0.2.8
- containers-0.5.10.2
- contravariant-1.4.1
- cookie-0.4.4
- cpphs-1.20.8
- cryptohash-0.11.9
- cryptohash-sha256-0.11.101.0
- cryptonite-0.25
- cryptonite-conduit-0.2.2
- data-default-class-0.1.2.0
- deepseq-1.4.3.0
- digest-0.0.1.2
- directory-1.3.0.2
- distributive-0.5.3
- dlist-0.8.0.4
- easy-file-0.2.1
- echo-0.1.3
- ed25519-0.0.5.0
- either-5
- exceptions-0.8.3
- extra-1.6.6
- fail-4.9.0.0
- fast-logger-2.4.11
- file-embed-0.0.10.1
- filelock-0.1.1.2
- filepath-1.4.1.2
- foundation-0.0.20
- free-5.0.1
- fsnotify-0.2.1.1
- generic-deriving-1.12.1
- ghc-boot-th-8.2.2
- ghc-prim-0.5.1.1
- gitrev-1.3.1
- hackage-security-0.5.3.0
- hashable-1.2.7.0
- haskell-src-exts-1.20.2
- haskell-src-meta-0.8.0.2
- hinotify-0.3.9
- hourglass-0.2.11
- hpack-0.28.2
- hpc-0.6.0.3
- hspec-2.4.8
- hspec-core-2.4.8
- hspec-discover-2.4.8
- hspec-expectations-0.8.2
- hspec-smallcheck-0.5.0
- http-api-data-0.3.7.2
- http-client-0.5.12.1
- http-client-tls-0.3.5.3
- http-conduit-2.3.1
- http-types-0.12.1
- integer-gmp-1.0.1.0
- integer-logarithms-1.0.2.1
- lifted-base-0.2.3.12
- logict-0.6.0.2
- memory-0.14.16
- microlens-0.4.8.3
- microlens-th-0.4.1.3
- mime-types-0.1.0.7
- mintty-0.1.1
- monad-control-1.0.2.3
- monad-logger-0.3.28.5
- monad-loops-0.4.3
- mono-traversable-1.0.8.1
- mtl-2.2.2
- mustache-2.3.0
- neat-interpolation-0.3.2.1
- network-2.6.3.5
- network-uri-2.6.1.0
- old-locale-1.0.0.7
- old-time-1.1.0.3
- open-browser-0.2.1.0
- optparse-applicative-0.14.2.0
- optparse-simple-0.1.0
- parsec-3.1.13.0
- path-0.6.1
- path-io-1.3.3
- path-pieces-0.2.1
- pem-0.2.4
- persistent-2.8.2
- persistent-sqlite-2.8.1.2
- persistent-template-2.5.4
- polyparse-1.12
- pretty-1.1.3.3
- primitive-0.6.3.0
- process-1.6.1.0
- profunctors-5.2.2
- project-template-0.2.0.1
- quickcheck-io-0.2.0
- random-1.1
- regex-applicative-0.3.3
- regex-applicative-text-0.1.0.1
- resource-pool-0.2.3.2
- resourcet-1.2.1
- retry-0.7.6.2
- rio-0.1.1.0
- rts-1.0
- safe-0.3.17
- scientific-0.3.5.3
- semigroupoids-5.2.2
- semigroups-0.18.4
- setenv-0.1.1.3
- silently-1.2.5
- smallcheck-1.1.3.1
- socks-0.5.6
- split-0.2.3.3
- stm-2.4.5.0
- stm-chans-3.0.0.4
- store-0.4.3.2
- store-core-0.4.1
- streaming-commons-0.1.19
- syb-0.7
- tagged-0.8.5
- tar-0.5.1.0
- template-haskell-2.12.0.0
- temporary-1.2.1.1
- text-1.2.3.0
- text-metrics-0.3.0
- tf-random-0.5
- th-abstraction-0.2.6.0
- th-expand-syns-0.4.4.0
- th-lift-0.7.8
- th-lift-instances-0.1.11
- th-orphans-0.13.5
- th-reify-many-0.1.8
- th-utilities-0.2.0.1
- time-1.8.0.2
- time-locale-compat-0.1.1.4
- tls-1.4.1
- transformers-0.5.2.0
- transformers-base-0.4.4
- transformers-compat-0.5.1.4
- typed-process-0.2.2.0
- unicode-transforms-0.3.4
- unix-2.7.2.2
- unix-compat-0.5.0.1
- unix-time-0.3.8
- unliftio-0.2.6.0
- unliftio-core-0.1.1.0
- unordered-containers-0.2.9.0
- uri-bytestring-0.3.1.1
- uuid-types-1.0.3
- vector-0.12.0.1
- vector-algorithms-0.7.0.1
- void-0.7.2
- x509-1.7.3
- x509-store-1.6.6
- x509-system-1.6.6
- x509-validation-1.6.10
- yaml-0.8.29
- zip-archive-0.3.2.4
- zlib-0.6.2

Warning: this is an unsupported build that may use different versions of
dependencies and GHC than the officially released binaries, and therefore may
not behave identically.  If you encounter problems, please try the latest
official build by running 'stack upgrade --force-download'.
```

### PR DESCRIPTION
**Pull request by rdnetto**
This was introduced in #3938, but breaks Docker mode under NixOS.

This PR relaxes the logic to only change the default under NixOS, rather than force Nix mode to enabled.

Fixes #4002.

Please include the following checklist in your PR:

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

Tested under NixOS using the Stack project.

### CODE REVIEW

**Comment by rdnetto** [comment]
@MartinPotier would you mind verifying that this fixes the problem? I don't have a good test case for using Docker under NixOS, though AFAICT using `--docker --no-nix` should just work.

**Comment by MartinPotier** [comment]
Yes, sure, let me find out how to do that and I'll get back to you asap

**Comment by MartinPotier** [comment]
This branch of stack does not build for me :/
I'm using LTS 11.8
```
[ 47 of 125] Compiling Stack.Types.VersionIntervals ( src/Stack/Types/VersionIntervals.hs, dist/build/Stack/Types/VersionIntervals.o )
src/Stack/Types/VersionIntervals.hs:57:3: error:
    • Couldn't match expected type ‘C.VersionIntervals’
                  with actual type ‘Maybe C.VersionIntervals’
    • In the expression: C.mkVersionIntervals $ map go vi
      In an equation for ‘toCabal’:
          toCabal (VersionIntervals vi)
            = C.mkVersionIntervals $ map go vi
            where
                go (VersionInterval lowerV lowerB mupper)
                  = (C.LowerBound (toCabalVersion lowerV) (toCabalBound lowerB),
                     case mupper of
                       Nothing -> C.NoUpperBound
                       Just (v, b) -> C.UpperBound (toCabalVersion v) (toCabalBound b))
   |
57 |   C.mkVersionIntervals $ map go vi
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
builder for '/nix/store/sffr9am9par2i5jpla1jvvliy4rsa2w6-stack-1.8.0.drv' failed with exit code 1
error: build of '/nix/store/sffr9am9par2i5jpla1jvvliy4rsa2w6-stack-1.8.0.drv' failed
```

**Comment by rdnetto** [comment]
Odd, that isn't a file I touched. CI is red for a different reason, which is also code I haven't touched.

What about if you just use `stack build`?

**Comment by MartinPotier** [comment]
Oh, right :D I guess that'd work

**Comment by MartinPotier** [comment]
Current stack
```
% stack --no-nix --docker --docker-image fpco/stack-build:lts-9.1 install --local-bin-path=./dist/bin --pedantic --flag mysoft1:production --flag mysoft2:production mysoft2
Note: Enabling Nix integration, as it is required under NixOS
Received ExitFailure 1 when running
Raw command: /run/current-system/sw/bin/docker run -v /nix/store/5c2mdypkz7wm959lvkmnwlm0gh6ayfn1-stack-1.7.1/bin/stack:/tmp/stack sha256:4e3c147fca483b64eedc025795f373f2aaba3559b1d23eafd6ac47dd638fc1ce /tmp/stack --version
```
Your version of stack
```
% /tmp/stack/.stack-work/install/x86_64-linux-nix/lts-11.6/8.2.2/bin/stack --no-nix --docker --docker-image fpco/stack-build:lts-9.1 install --local-bin-path=./dist/bin --pedantic --flag mysoft1:production --flag mysoft2:production mysoft2
Received ExitFailure 1 when running
Raw command: /run/current-system/sw/bin/docker run -v /tmp/stack/.stack-work/install/x86_64-linux-nix/lts-11.6/8.2.2/bin/stack:/tmp/stack sha256:4e3c147fca483b64eedc025795f373f2aaba3559b1d23eafd6ac47dd638fc1ce /tmp/stack --version
```

Although, with stack 1.6.5 it proceeds fine. I guess it wasn't the problem, or the only problem…

**Comment by rdnetto** [comment]
I think I know what's happening - run `ldd` on the Stack binary inside the container to confirm.
My hunch is that the Stack built outside the container links against a different libc (or some other library) to what's present inside the container, which is a fairly common problem with NixOS since it doesn't have its libraries in the standard locations.

You should be able to fix it by using a Docker container that matches the host environment.

**Comment by MartinPotier** [comment]
Well…
```
% /run/current-system/sw/bin/docker run -v /nix/store/5c2mdypkz7wm959lvkmnwlm0gh6ayfn1-stack-1.7.1/bin/stack:/tmp/stack sha256:4e3c147fca483b64eedc025795f373f2aaba3559b1d23eafd6ac47dd638fc1ce /usr/bin/ldd /tmp/stack
/tmp/stack: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /tmp/stack)
        linux-vdso.so.1 =>  (0x00007fff5b7b8000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe46f4e0000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fe46f2c6000)
        libsqlite3.so.0 => /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007fe46eff1000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe46edd4000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fe46ebcc000)
        libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fe46e9c9000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe46e7c5000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007fe46e545000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe46e17b000)
        /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007fe46f7e9000)
```
and
```
% ldd /nix/store/5c2mdypkz7wm959lvkmnwlm0gh6ayfn1-stack-1.7.1/bin/stack
        linux-vdso.so.1 (0x00007ffe469ba000)
        libm.so.6 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libm.so.6 (0x00007fa57fe2c000)
        libz.so.1 => /nix/store/vj9716h62y21a2hv18al8zvhlqpl30sx-zlib-1.2.11/lib/libz.so.1 (0x00007fa57fc15000)
        libsqlite3.so.0 => /nix/store/kbfvlkq41zcz4vmgiimvp59m9nxhgi5q-sqlite-3.22.0/lib/libsqlite3.so.0 (0x00007fa57f914000)
        libpthread.so.0 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libpthread.so.0 (0x00007fa57f6f5000)
        librt.so.1 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/librt.so.1 (0x00007fa57f4ed000)
        libutil.so.1 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libutil.so.1 (0x00007fa57f2ea000)
        libdl.so.2 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libdl.so.2 (0x00007fa57f0e6000)
        libgmp.so.10 => /nix/store/hwkkw10kb5w600c3dvv061gn1ld5fbja-gmp-6.1.2/lib/libgmp.so.10 (0x00007fa57ee52000)
        libc.so.6 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libc.so.6 (0x00007fa57ea9e000)
        /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/ld-linux-x86-64.so.2 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007fa5801c1000)
```
are the same, aren't they?

**Comment by rdnetto** [comment]
Rebased to resolve conflicts.

**Comment by MartinPotier** [comment]
Is stack really depending on the contained system to be running the same libraries than the host? (agreed it's another topic)

File: stack.yaml [hunk 1]
```diff
@@ -9,7 +9,6 @@ resolver: lts-11.6
 #       name: "fpco/stack-test"
 nix:
   # --nix on the command-line to enable.
-  enable: false
```
Comment (snoyberg) [inline stack.yaml]: What's the motivation for this change?

Comment (rdnetto) [reply stack.yaml]: Before, `enable: false` was the default, so this line had no effect. After this PR is applied, the default value depends on whether we're running under NixOS or not, so it's desirable to omit it to build correctly on both NixOS and other distros.

Comment (snoyberg) [reply stack.yaml]: Makes sense, thanks!

**Comment by ip1981** [comment]
It is not just Docker. I run stack on NixOS and don't want it support of Nix at all.

**Pull request merged**